x-pack/plugins/fleet/server/services/agent_policy.ts

```diff
@@ -715,31 +723,91 @@ class AgentPolicyService {
-      return;
-    }
+    const policies = await agentPolicyService.getByIDs(soClient, agentPolicyIds);
+    const policiesMap = keyBy(policies, 'id');
+    const fullPolicies = await Promise.all(
+      agentPolicyIds.map((agentPolicyId) =>
+        // There are some potential performance concerns around using `getFullAgentPolicy` in this context, e.g.
+        // re-fetching outputs, settings, and upgrade download source URI data for each policy. This could potentially
+        // be a bottleneck in environments with several thousand agent policies being deployed here.
+        agentPolicyService.getFullAgentPolicy(soClient, agentPolicyId)
+      )
+    );
 
-    const fleetServerPolicy: FleetServerPolicy = {
-      '@timestamp': new Date().toISOString(),
-      revision_idx: fullPolicy.revision,
-      coordinator_idx: 0,
-      data: fullPolicy as unknown as FleetServerPolicy['data'],
-      policy_id: fullPolicy.id,
-      default_fleet_server: policy.is_default_fleet_server === true,
-    };
+    const fleetServerPolicies = fullPolicies.reduce((acc, fullPolicy) => {
+      if (!fullPolicy || !fullPolicy.revision) {
+        return acc;
+      }
 
-    if (policy.unenroll_timeout) {
-      fleetServerPolicy.unenroll_timeout = policy.unenroll_timeout;
-    }
+      const policy = policiesMap[fullPolicy.id];
+      if (!policy) {
+        return acc;
+      }
+
+      const fleetServerPolicy: FleetServerPolicy = {
+        '@timestamp': new Date().toISOString(),
+        revision_idx: fullPolicy.revision,
+        coordinator_idx: 0,
+        data: fullPolicy as unknown as FleetServerPolicy['data'],
+        policy_id: fullPolicy.id,
+        default_fleet_server: policy.is_default_fleet_server === true,
+      };
+
+      if (policy.unenroll_timeout) {
+        fleetServerPolicy.unenroll_timeout = policy.unenroll_timeout;
+      }
 
-    await esClient.create({
+      return [...acc, fleetServerPolicy];
+    }, [] as FleetServerPolicy[]);
+
+    const fleetServerPoliciesBulkBody = fleetServerPolicies.flatMap((fleetServerPolicy) => [
+      {
+        index: {
+          _id: uuidv5(
+            `${fleetServerPolicy.policy_id}:${fleetServerPolicy.revision_idx}`,
+            uuidv5.DNS
+          ),
+        },
+      },
+      fleetServerPolicy,
+    ]);
+
+    const bulkResponse = await esClient.bulk({
       index: AGENT_POLICY_INDEX,
-      body: fleetServerPolicy,
-      id: uuid(),
+      body: fleetServerPoliciesBulkBody,
       refresh: 'wait_for',
     });
+
+    if (bulkResponse.errors) {
+      const logger = appContextService.getLogger();
+      const erroredDocuments = bulkResponse.items.reduce((acc, item) => {
+        const value: BulkResponseItem | undefined = item.index;
+        if (!value || !value.error) {
+          return acc;
+        }
+
+        return [...acc, value];
+      }, [] as BulkResponseItem[]);
+
+      logger.debug(`deployPolicies failed to index: ${JSON.stringify(erroredDocuments)}`);
```

Comment:
Hmm, super nitpicky here, but could we avoid having the method name in this debug log and lean towards "plain english" logging output? e.g.

```ts
logger.debug(`Failed to index documents during policy deployment: ${JSON.stringify(erroredDocuments)}`)
```